cryostatio/cryostat-operator
# feat(agent): add pod mutation webhook to inject agent

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: #970 

## Description of the change:
* Adds a mutating admission webhook for pods that responds to a pair of labels by injecting the Cryostat agent and configuring it to work with the specified Cryostat instance.

## Motivation for the change:
* Makes it much easier to set up the Cryostat agent with arbitrary pods.

## How to manually test:
1. `make deploy`
2. `make create_cryostat_cr`
3. `make sample_app`
4. Modify the sample app deployment as follows:
    1. Override `JAVA_OPTS_APPEND` to not inject the built-in Cryostat agent.
        ```
        - env:
          - name: JAVA_OPTS_APPEND
            value: -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -Dio.cryostat.agent.shaded.org.slf4j.simpleLogger.defaultLogLevel=debug
        ```
    2. Add the following labels:
        ```
        cryostat.io/name: cryostat-sample
        cryostat.io/namespace: cryostat-operator-system
        ```
    3. I had to also increase the memory limit to prevent OOM.
5. The sample application should appear as an agent target in Cryostat

I've filed issues for the TODOs in this PR. There are a few of them, but they're all pretty small compared to this PR:
#992, #993, #994, #995, #996
